### PR TITLE
Added Several New Reference Events

### PR DESCRIPTION
**Pull request by addisonbeck**
### Related Work
[Asana ticket](https://app.asana.com/0/1183359552741420/1200418300903249/f)

### Code Changes
* Implemented or tweaked several reference events. Following along commit by commit is is a good way to read through this.
* [Refactor]: I created some static string and assigned them to clients in the client store so I could reuse that information for  making sure the directorySync event only fires when connected to the directoryConnector. This also involved adding client information to current context.

### Note
Two events from the requirements are not implemented yet: `trialStart` and `trialEnd`. It is still unclear how we are going to fire those events. A big part of trial data involves the process sales uses to create and end trials for clients manually, as they don't seem to do that in the admin portal. I requested info from product about this, but didn't want to leave these working events in the ether in the meantime.

### CODE REVIEW

File: src/Core/Enums/ReferenceEventType.cs [hunk 1]
```diff
@@ -28,5 +28,17 @@ public enum ReferenceEventType
         SendCreated,
         [EnumMember(Value = "send-accessed")]
         SendAccessed,
+        [EnumMember(Value = "organization-imported")]
+        DirectorySynced,
+        [EnumMember(Value = "vault-imported")]
+        VaultImported,
+        [EnumMember(Value = "first-secret-added")]
```
Comment (cscharf) [inline src/Core/Enums/ReferenceEventType.cs:35]: Please remove "first-" from enum values/members and just leave them `SecretAdded`, `GroupCreated`, `CollectionCreated`

File: src/Core/Services/Implementations/CipherService.cs [hunk 2]
```diff
@@ -83,6 +87,12 @@ public class CipherService : ICipherService
                         cipher.UserId = savingUserId;
                     }
                     await _cipherRepository.CreateAsync(cipher, collectionIds);
+
+                    if ((await _cipherRepository.GetManyByOrganizationIdAsync(cipher.OrganizationId.Value)).Count < 1)
```
Comment (kspearrin) [inline src/Core/Services/Implementations/CipherService.cs:91]: See my comments on Asana. We don't want to have to do these calculations.

**Review by kspearrin — CHANGES_REQUESTED** [review]

File: src/Core/Services/Implementations/CipherService.cs (hunk 2 above)
Comment (cscharf) [inline src/Core/Services/Implementations/CipherService.cs:91]: Remove the `if` statement, just raise the event each time.

File: src/Core/Services/Implementations/CollectionService.cs [hunk 3]
```diff
@@ -76,6 +81,10 @@ public class CollectionService : ICollectionService
                 }
 
                 await _eventService.LogCollectionEventAsync(collection, Enums.EventType.Collection_Created);
+                if ((await _collectionRepository.GetManyByOrganizationIdAsync(org.Id)).Count.Equals(1))
```
Comment (cscharf) [inline src/Core/Services/Implementations/CollectionService.cs:84]: Remove the `if` statement and always raise the event.

File: src/Core/Services/Implementations/GroupService.cs [hunk 4]
```diff
@@ -54,6 +59,10 @@ public async Task SaveAsync(Group group, IEnumerable<SelectionReadOnly> collecti
                 }
 
                 await _eventService.LogGroupEventAsync(group, Enums.EventType.Group_Created);
+                if ((await _groupRepository.GetManyByOrganizationIdAsync(org.Id)).Count.Equals(1))
```
Comment (cscharf) [inline src/Core/Services/Implementations/GroupService.cs:62]: remove `if` block and always raise the event.

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 5]
```diff
@@ -1243,7 +1246,8 @@ public async Task DisableTwoFactorProviderAsync(Organization organization, TwoFa
             await _referenceEventService.RaiseEventAsync(
                 new ReferenceEvent(ReferenceEventType.InvitedUsers, organization)
                 {
-                    Users = orgUserInvitedCount
+                    Users = orgUserInvitedCount,
+                    FirstInvite = userCount < 2
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1250]: You can remove this.

File: src/Core/Services/Implementations/OrganizationService.cs [hunk 6]
```diff
@@ -1175,9 +1178,9 @@ public async Task DisableTwoFactorProviderAsync(Organization organization, TwoFa
                 await ValidateOrganizationUserUpdatePermissions(organizationId, invite.Type.Value, null);
             }
 
+            var userCount = await _organizationUserRepository.GetCountByOrganizationIdAsync(organizationId);
```
Comment (cscharf) [inline src/Core/Services/Implementations/OrganizationService.cs:1181]: You can move this back into context (undo this and the next change).

File: src/Core/Models/Business/ReferenceEvent.cs [hunk 7]
```diff
@@ -52,5 +57,7 @@ public ReferenceEvent(ReferenceEventType type, IReferenceable source)
         public int? MaxAccessCount { get; set; }
 
         public bool? HasPassword { get; set; }
+
+        public bool? FirstInvite { get; set; }
```
Comment (cscharf) [inline src/Core/Models/Business/ReferenceEvent.cs:61]: Please remove this property, it's not necessary.

**Review by cscharf — DISMISSED** [review]

**Comment by cscharf** [comment]
Looks like @kspearrin needs to perform a final review.

File: src/Core/Enums/ReferenceEventType.cs [hunk 8]
```diff
@@ -28,5 +28,17 @@ public enum ReferenceEventType
         SendCreated,
         [EnumMember(Value = "send-accessed")]
         SendAccessed,
+        [EnumMember(Value = "organization-imported")]
+        DirectorySynced,
+        [EnumMember(Value = "vault-imported")]
+        VaultImported,
+        [EnumMember(Value = "secret-added")]
+        SecretAdded,
```
Comment (kspearrin) [inline src/Core/Enums/ReferenceEventType.cs:36]: Should this be called `CipherCreated` to be consistent with codebase terminology?

**Review by kspearrin — APPROVED** [review]

**Pull request merged**